### DISCUSSION TITLE
use of run-specific q-value filter instead of global q-value filterin

### DISCUSSION THREAD

**Discussion by JM-Bader** [vdemichev/DiaNN#493] [Q&A] [answered]
I process a large cohort of thousands of samples with the DIANN GUI 
It appears that this leads to a problem with the global FDR filtering as random decoys accumulate over the thousands of raw files and lead to significant FDR-based cutting of the final dataset. 

Q1: Is it possible to only apply a run-specific FDR and disable the global FDR?

Q2: I then plan to apply a valid value filter to remove the excess of decoy 'one-hit wonders' (and their corresponding 'one-hit wonder' wrong targets). I guess that makes sense, doesn't it?

Q3: The github suggests that a run-specific FDR instead of global FDR is possible but I don't really get how to set this in the GUI with commands. How do I do it?

Q4: Should the global FDR-filtering be disabled on precursor, peptide, or protein level ? Or all of them (as accumulation of random decoys over thousands of runs can happen in all these categories)?

**Comment by vdemichev** [comment, marked as answer]
Hi Jakob,

What you observe is strange. Do you mean you get A LOT less precursors at 1% global FDR with 5000 runs than with, say, 100 runs? There are algorithmic measures in DIA-NN 1.8 and later specifically designed to make sure that this does not happen. How do the numbers you observe look like?

Q1: Without MBR there's no global filtering applied to the main report. With MBR you can relax it a bit indeed. To, say, 2% or 3% should be OKish.
Q2: What do you mean 'decoy' one-hit wonders? DIA-NN does not report decoy hits at all.
Q3: Change precursor FDR margin in the GUI to 2%, for example. This will set MBR global FDR threshold to 2%. Run-specific will also be 2% in the final report, but that you can always filter to 1% afterwards.
Q4: I would never disable it fully. Relaxing protein global FDR a bit for plasma might make sense. For tissues I don't think this is ever necessary.

Best,
Vadim

**Reply by JM-Bader** [reply]
I noticed it at protein level first. 
When searching about 500 files, I get 2732 protein groups
When searching about 7000 files, I get 1821 protein groups
it is the identical library

**Reply by vdemichev** [reply]
Which filtering exactly?

**Reply by vdemichev** [reply]
Main report or pg_matrix? Numbers of precursors?

**Reply by JM-Bader** [reply]
Q1: that is interesting. the library is the result of a DIANN search with MBR. However, the actual searches (500 vs 7000) produce more or less equivalent results with (second pass) or without MBR (first pass), i.e the reduced number of proteins with large dataset occurs regardless of MBR. 

Q2: Sorry, just for explanatory purposes. MaxQuant does report decoys (and labels them as decoys) so for a moment I confused it... I was trying to make the point that accumulating wrong 'one hit wonder' targets will be likewise removed.

**Reply by JM-Bader** [reply]
filtering: no filtering, reported protein groups (including those with low data completeness)
proteins: main report-extracted proteins
Re precursor I will have to analyse the data

**Reply by vdemichev** [reply]
Q1: OK, this explains it. Please see the docs on this: with MBR it's Lib.Q.Value not Global.Q.Value that should be used. If you manually supply a DIA-derived library to DIA-NN, it does not know that you are 'reproducing' MBR, and filters using the more srict global q-value. So with DIA-derived lib, you need only (i) run-specific precursor (ii) Lib.PG.Q.Value - manual fitlering using it and (iii) PG.Q.Value if you feel like you need it.

**Reply by vdemichev** [reply]
Also, in general don't want to use MBR with highly-specific libs, like most DIA-derived libs are.

**Reply by JM-Bader** [reply]
Q4: I see your point and agree that this makes more sense for high dynamic range samples than for tissues. My case is CSF. I presume it is valid there as well (similar to plasma)

**Reply by JM-Bader** [reply]
> Also, in general don't want to use MBR with highly-specific libs, like most DIA-derived libs are.

I know, that's why I looked at first pass results. (however no difference between first pass and second pass results)

**Reply by JM-Bader** [reply]
> Q1: OK, this explains it. Please see the docs on this: with MBR it's Lib.Q.Value not Global.Q.Value that should be used. If you manually supply a DIA-derived library to DIA-NN, it does not know that you are 'reproducing' MBR, and filters using the more srict global q-value. So with DIA-derived lib, you need only (i) run-specific precursor (ii) Lib.PG.Q.Value - manual fitlering using it and (iii) PG.Q.Value if you feel like you need it.

So if I supply a DIA-derived lib to DIA-NN it is not sufficient to disable MBR but to use (i) run-specific precursor (ii) Lib.PG.Q.Value - manual fitlering using it and (iii) PG.Q.Value if I feel like Ineed it.

But what are the commands to do (i) ?
(ii) and (iii) are purely processing steps of the main report tsv, correct?

**Reply by JM-Bader** [reply]
Lib.PG.Q.Value is a column / parameter reported for all proteins in the main report tsv I presume, correct?

**Reply by JM-Bader** [reply]
![image](https://user-images.githubusercontent.com/98888664/188886180-efa0a1be-942d-4a8d-85e9-b59d07b0866b.png)

this is the section in the docs that appears most likely relevant. 
P1) --matrix-spec-q appears to be almost the right thing. However it says 'in addition to global q-value filtering'. But I need to use the run-specific q-value _**_instead_**_ of the global q-value filtering, correct? 
P2) What are 'protein matrices' ? This is ambiguous as --matrices enables all the output files (e.g pg.matrix) on top of the main report. So do these commands affect only the pg.matrix table and alike but not the main report ? or all output formats?
P3) What is a translated q-value?

**Reply by JM-Bader** [reply]
Anyway, I now tested a search with --matrix-spec-q in the GUI (and the library being the result of a DIA-NN search). I get even fewer proteins which aligns with the github description (saying that the run-specific q-value is applied on top of the global one - making it more stringent than the global one alone). In effect, --matrix-spec-q appears not to be right command and it is unclear to me what the right command is

**Reply by vdemichev** [reply]
I would suggest to exclusively look at the main report, not the matrices.

**Reply by vdemichev** [reply]
P1: docs are correct, --matrix-spec-q applies extra filtering. Which means whatever the setting it will be too strict filtering if you use DIA-derived lib.

**Reply by vdemichev** [reply]
P2: only the files with 'matrix' in their name

**Reply by vdemichev** [reply]
P3: it's for multiplexing, see info on the --channels

**Reply by JM-Bader** [reply]
> I would suggest to exclusively look at the main report, not the matrices.

I do.

**Reply by JM-Bader** [reply]
> 



> Q1: OK, this explains it. Please see the docs on this: with MBR it's Lib.Q.Value not Global.Q.Value that should be used. If you manually supply a DIA-derived library to DIA-NN, it does not know that you are 'reproducing' MBR, and filters using the more srict global q-value. So with DIA-derived lib, you need only (i) run-specific precursor (ii) Lib.PG.Q.Value - manual fitlering using it and (iii) PG.Q.Value if you feel like you need it.

So what is the right command to execute that search (i) with run-specific q-values only (and no global q-value filter for the main report)?

**Reply by vdemichev** [reply]
Main report is never filtered at global q-value. So can use it right away

**Reply by JM-Bader** [reply]
> Main report is never filtered at global q-value. So can use it right away

That is sursprising. 
So precursors that are not filtered for a global q-value contribute to protein intensities assembled by DIA-NN?

**Reply by JM-Bader** [reply]
and by default the global precursor q-value is just reported for information but plays no role at all?

**Reply by JM-Bader** [reply]
Moreover, if there is no command to change the filtering of precursors from 'Global.Q.Value' to 'Lib.Q.Value' and I do this manually by filtering the main report, how do I get the updated protein group intensities when changing the precursor filtering?

**Reply by JM-Bader** [reply]
There is no 'PG.MaxLFQ_Based on Global.Q.Value' and 'PG.MaxLFQ_Based on Lib.Q.Value' or equivalently in the main report as far as I can tell...

**Reply by vdemichev** [reply]
Yes, not-filtered precursors contribute to protein intensities. MaxLFQ is fairly robust to this. And one can always use MaxLFQ manually in R using anhy set of peptides. But if you have a DIA library derived from the same experiment or use MBR this means the global FDR filtering is unnecessary, as the library is already filtered.

**Reply by vdemichev** [reply]
Global precursors q-value is reported so can be used for filtering if the user needs it for the specific workflow.

**Reply by vdemichev** [reply]
To update protein groups intensities, please use MaxLFQ as implemented in the 'iq' or 'diann' R packages. iq is a lot faster

**Reply by JM-Bader** [reply]
I see. 
So far I have just compared protein groups across searches - each taken from the main report without filtering (since I assumed these protein group entries are valid in the sense that may be used for analysis). 
Q1: Is this the case, are the protein group entries in the main report valid without filtering?
Q2: Comparing the DIA-NN output library (+MBR)-based re-searches vs MSFragger-based searches, I also didn't filter proteins for any q-value in the main report. This caused the phenomenon initially reported. If I now 'not filter on Global.Q.Value' but on 'Lib.Q.Value' in the DIA-NN output library (+MBR)-based re-searches, this does not increase the number of proteins in these searches because initially I didn't filter at all...  > So the phenomenon still persists!
So is the conclusion that many proteins weren't supposed to be used any way, because I didn't filter?
... And when I filter both search types, I will get fewer proteins overall but the phenomenon (comparatively fewer proteins in DIA-NN Lib re-searches is at least gone) ?

**Reply by vdemichev** [reply]
Q1: they are not filtered at any protein FDR.

As I understand, you do the following?
- Make a DIA-based library using lib-free search of a subset of your runs
- Analyse with this library the whole dataset (thousands of runs)

Then all you need to do is filter the main report using Lib.PG.Q.Value, with reasonable thresholds being 1%-3%. Further, PG.Q.Value filtering is optional (will reduce data completeness). No need to recalculate protein quantities. So all is very simple :)

**Reply by JM-Bader** [reply]
yes. 
- the DIA-based library is based on a search with MBR. 
- the final search (thousands of runs) also has MBR activated but I use the first pass main report output

**Reply by JM-Bader** [reply]
If I want to compare the results with a 'MSFragger lib-based direct search' (search #1) with this strategey, 'DIA-NN library free search-based library' (search #2), I 
- filter the main report of search#1 for 1% Global.PG.Q.Value (a protein-level filter)
- filter the main report of search#2 for 1% Lib.PG.Q.Value (a protein level filter)

for a fair comparison?

**Reply by JM-Bader** [reply]
and no other filtering (e.g. precursor level) other than this single protein-level filtering to get sound quantitative data usable for downstream analysis on protein level?

**Reply by vdemichev** [reply]
The filtering in DIA-NN and FragPipe + DIA-NN DIA workflows will be different on the protein level, it's impossible to easily make it equal. So cannot really compare directly. The only thing that can compare is precursor numbers filtered at 1% run-specific FDR. Everything else is incomparable.

**Reply by JM-Bader** [reply]
Q1: How about running the library-free search to generate the library without MBR and turning MBR on for the large search of the full dataset?

**Reply by JM-Bader** [reply]
Q2: ok, independently then: If I want to get protein group intensities for downstream analysis from the main report of a regular fragpipe library, how do I have to filter the main report? 1% Global.PG.Q.Value ? Is this how the pg.matrix table is generated from the main report?

**Reply by vdemichev** [reply]
Q1: Does not make sense to turn MBR on. MBR only helps with libraries that are not highly specific to the project.
Q2: Global.PG.Q.Value, yes. FragPipe libraries don't have Lib. q-values.

**Reply by JM-Bader** [reply]
Feedback: Filtering the main report table of a search based on a DIA-NN output library by Lib.PG.Q.Value produces results that look reasonable to me so far. This is a cool strategy. Making it more apparent/explicit in the docs may be helpful.
You explained that the Lib.PG.Q.Value filter at a reasonable level (e.g. 1-3%) should be sufficient to filter the main report. Overall, I think it is. I observer a very few entries that are left with quite high Global.PG.Q.Value: 
![image](https://user-images.githubusercontent.com/98888664/189307401-5d820e21-f175-46a2-9995-27f87979676c.png)
So I think having a relaxed Global.PG.Q.Value filter on top, e.g. < 10% is not stringent but removes some very suspicious entries.

In any case, when trying to publish said dataset with that filtering strategy, is there paper to reference ? It may be strange to reviewers that the data is not  or not stringently filtered by Global.PG.Q.Value (as this normally the case).

**Reply by vdemichev** [reply]
Global.PG.Q.Value just does not make sense, if filtering with Lib.PG.Q.Value. Yes, it can be that those proteins are actually fine, but because Global.PG.Q.Value is too stringent, it's equal to 1 for them...

Lib.PG.Q.Value is literally the 'global protein q-value' for the analysis which was used to create the library. So if you create it from a subset of your dataset, that's what needs to be used.

Something I forgot to mention, which I guess I should add to the docs. In your scenario, unlike when using MBR, protein grouping will be carried out anew, if enabled. As you'd like to filter based on Lib.PG.Q.Value instead of Global.PG.Q.Value, you don't want that. So just replace the contents of the UniprotID column in the library with the contents of the Protein.Group column and disable protein inference in DIA-NN when analysing with this library. Then it will use the protein groups generated during the library generation step.

**Reply by vdemichev** [reply]
Alternatively, `--library-headers *,*,*,*,*,*,Protein.Group` (lets DIA-NN know to use the Protein.Group column as the main protein ID column) might also work instead of editing the library file, but I haven't tested this for a while.

**Reply by JM-Bader** [reply]
thanks for the information. 


> Global.PG.Q.Value just does not make sense, if filtering with Lib.PG.Q.Value. Yes, it can be that those proteins are actually fine, but because Global.PG.Q.Value is too stringent, it's equal to 1 for them...
> 
> Lib.PG.Q.Value is literally the 'global protein q-value' for the analysis which was used to create the library. So if you create it from a subset of your dataset, that's what needs to be used.
> 
> Something I forgot to mention, which I guess I should add to the docs. In your scenario, unlike when using MBR, protein grouping will be carried out anew, if enabled. As you'd like to filter based on Lib.PG.Q.Value instead of Global.PG.Q.Value, you don't want that. So just replace the contents of the UniprotID column in the library with the contents of the Protein.Group column and disable protein inference in DIA-NN when analysing with this library. Then it will use the protein groups generated during the library generation step.

thanks for the information. 
What about just disabling protein grouping by setting protein inference to 'Off'. Wouldn't that be sufficient?

**Reply by vdemichev** [reply]
If you just set the inference off, DIA-NN will report all protein IDs from the UniprotID column of the library, which is all the proteins associated with the peptide, without any grouping

**Reply by JM-Bader** [reply]
> If you just set the inference off, DIA-NN will report all protein IDs from the UniprotID column of the library, which is all the proteins associated with the peptide, without any grouping

Q1: ah. Do I understand correctly that generally: using 'protein inference off' in combination with a library leads to reporting of all proteins (as of UniProt IDs) without any protein grouping instead of using the protein grouping present in the library ? (independently of whether the library is DIA-NN-derived or other?)
... But then people should find solutions for that often (if they want to preserve Lib protein grouping), not just in this scenario here.

**Reply by JM-Bader** [reply]
> 



> Alternatively, `--library-headers *,*,*,*,*,*,Protein.Group` (lets DIA-NN know to use the Protein.Group column as the main protein ID column) might also work instead of editing the library file, but I haven't tested this for a while.

I will try this. 
Q2: Do I need to turn protein inference off as well here ? 
Q3: Unless it produces an error, what is an efficient way to check whether it worked or not? / How do I know it worked?

**Reply by vdemichev** [reply]
Q1: protein inference off = no protein inference performed by DIA-NN. So reports whatever is in the respective library column.
Q2: yes
Q3: test on 2 runs, every protein group in the output would be identical to the respective protein group for that precusor in the librayr

**Reply by JM-Bader** [reply]
Q1: 'respective library column' means it looks for the UniProt_ID column in the library (not the Protein.Group column in the library), correct? Wouldn't be sensible to generally change that ?

**Reply by vdemichev** [reply]
Q1: Correct

**Reply by JM-Bader** [reply]
Ok, I understand. As said it appears to me that this behavior is generally unwanted and could/should thus be generally changed. Or am I missing sth.

**Reply by JM-Bader** [reply]
or maybe make that a button in the GUI to select library grouping instead of UniProt ID

**Reply by JM-Bader** [reply]
> Alternatively, `--library-headers *,*,*,*,*,*,Protein.Group` (lets DIA-NN know to use the Protein.Group column as the main protein ID column) might also work instead of editing the library file, but I haven't tested this for a while.

FYI: this command causes DIANN not to recognize proteins in the library properly (see top row for command, bottom for issue): 
![image](https://user-images.githubusercontent.com/98888664/189672394-34c85da0-d9b6-43a6-b148-7254a56008f7.png)

**Reply by JM-Bader** [reply]
Q2: FYI, copying the ProteinGroup column from the DIA-NN-generated library to overwrite the UniprotID column there expectedly works - in the sense that DIANN is able to read the file, and accordingly adjusts the number of proteins in the library. I set Protein inference to OFF and unticked heuristic protein inference. 
Does it make a differnce if there is fasta file still loaded ? (For now I left it, so it is still specified). I guess at this point the fasta is irrelevant though. Correct?

**Reply by JM-Bader** [reply]
Q3:  Independently, for the Re-search I plan to use the first pass results. Does it make a difference if MBR is enabled (which also leads to a second pass and it's results) ? i guess it does not make a difference for first pass results. Correct?

**Reply by JM-Bader** [reply]
FYI as feedback: 
Correctly transforming the library (using Protein.Groups to overwrite UniprotID column in DIA-NN output lib) does lead to a noticeable difference (which is reasonable as well!):

blue: initial search using MSFragger lib
orange: re-search using output library of initial search (not transforming the lib)
green: re-search using output library of initial (transforming the lib) + Protein grouping disabled in the re-search
Calculation of numbers: filter main report tsv for 1% Global.PG.Q.Value (initial search) or 1% Lib.PG.Q.Value (re-searches). For initial search, take second pass results, for re-searches first pass results (i.e disabling MBR). 
![image](https://user-images.githubusercontent.com/98888664/189905457-234d1951-0714-4eab-bff2-45ee0892d8ec.png)

The green and blue overlap much more than the orange one. 
So this library transformation is actually important in my dataset

**Comment by JM-Bader** [comment]
Slightly different topic here: All of the above is about proteins. 
Here, how about precursors and fragments? (relevant to manual assembly of protein intensities with iQ etc.)

Q1: 
Like for proteins, I would now expect in the main report:
- run-specific q-value
- Global.Q.Value
- Lib.Q.Value
- each for precursors and fragments. 
Do I find these 3 q-values there each? 

Q2: 
I presume the table is already filtered for run-specific q-value each but not regarding the two other q-values - each for fragments and precursors. Correct?

Q3: 
Do I need to filter the main report before assembling protein intensities based on precursor and fragment information with iQ?
Should I use the Lib.Q.Val when the library is a DIA-NN output and the Global.Q.Value when the Library is not a DIA-NN output?

**Reply by vdemichev** [reply]
Q1: Why would you want to have fragment q-value? Precursor q-value indicates that the precursor is there. Of course, it does not guarantee that it's correctly quantified - this is something we can't do properly atm :)
Q2: Run-specific q-value is precursor-level
Q3: If using MBR - no. If using DIA-derived lib - based on Lib.Q.Value. Yes, exactly.

**Reply by JM-Bader** [reply]
Q1: 
a) for potential quality control. b) My idea was to filter for well-identified fragments and also use these for protein quantification (rather than all fragments). Or at least check whether that change (all vs well-identified fragments) correlates with a difference in protein CVs etc. I see your point that this would maximally go half way as the q-value controls for identification but not quantification and second your experience that MaxLFQ is relatively stable towards wrong precursors. 

a) So in essence, DIA-NN currently only uses precursor intensity for protein quantification ? 
b) DIA-NN filters precursors for run-specific q-value (but not global precursor Q-value and lib Q-value) for i) reporting in the main tsv and ii) protein quantification - correct?
c) .... So, what happens if fragments are identified and quantified but the precursor is not detected (e.g. in a crowded MS1 space) ? As far as I remember, this is actually not uncommon

Q3: I see, your answer pertains to precursors only as currently DIANN MaxLFQ only uses precurso intensity and you recommend to do it likewise with iQ?

**Reply by vdemichev** [reply]
Q1a Yes, makes sense :) But not implemented yet
a Yes
b Yes, but MBR means automatic global q-value (called Lib.Q.Value in this case) filtering
c Quant is MS2 based in DIA-NN
Q3 Yes, but it's been reported that fragment-level MaxLFQ also works nicely

**Reply by JM-Bader** [reply]
Wait, I am confused now. 
Q3-answer: DIA-NN uses MaxLFQ to quantify proteins based on precursor intensities
Q1c-answer: DIA-NN uses MS2 quantities (i.e. fragments, and not precursors) to quantify proteins
Or do you mean that in a first step precursor intensities are inferred from MS2 level intensities (=fragments) (rather than integrating the MS1 traces of precursors) ? And subsequently precursor intensities are fed into MaxLFQ?

**Reply by vdemichev** [reply]
Sorry, 'precursor' in DIA-NN means a charged peptide molecule species, which is quantified using MS/MS signal, i.e. fragment intensities

**Reply by JM-Bader** [reply]
ok, understood. I just incorrectly assumed that precursor intensity was inferred from MS1 traces of the precursors (which may have quite a lot missing values). Ok, things make sense now

**Comment by JM-Bader** [comment]
yet another aspect: 
if I run a primary search (i.e. using a MSFragger library), I get both the main report and the library from that search. 
Surprisingly, the library contains more unique Protein.Group entries than the main report filtered at 1% Global.PG.Q.Value. I then looked at the 'PGQValue' reported in the library file - which I presume is the Global.PG.Q.Value in the main report (Q1: correct?). Unexpectedly, the library contains entries with a PGQValue > 1%. 
Q2: Why is that (lib entries with PG.Q.value > 1%)? Is there another PGQValue filter (e.g. 10%) or none at all? What is the point of having them in the library? Using a very relaxed filtering in the secondary search?
Q3: When using said library for a Re-Search, these entries will be removed on protein level (when using a PG.Lib.Q.Value filter < 1%), so they don't matter anyway, do they ? Or is there any chance they can become significant (i.e. included in the final search results) ? Hypothetically, if protein grouping is changed the precursors  could be assigned to new groups - but that is not desired (because PG.Lib:Q.Value filtering then doesn't make sense anymore)... Or am I missing anything?
Q4: When comparing unique Protein.Group entries, the library still has more entries than the unfiltered (ie. not Global.PG.Q.Value-filtered) main report. Why?

**Reply by vdemichev** [reply]
MSFragger libs should be used with protein inference disabled in DIA-NN.
When DIA-NN generates a lib from DIA data, it's filtered using global precursor FDR. It's not filtered using any protein FDR as the user can always do such filtering after the analysis is finished.
Q3: sorry, got a bit lost about what's happening.

**Reply by JM-Bader** [reply]
Ok, I see the logic. 
Re Q3: I guess I don't have to care about the number of proteins in the library. I just use overwrite the UniProtID column with the PG.Group column there and remove all proteins in the final search which have a Lib.PG.Q.Value > 1%. I definitely need to disable protein grouping and heuristics for the re-search using the DIA-NN lib.

**Reply by vdemichev** [reply]
Yes
